Propose a draft TRD for app completion codes (#2914)
tock/tock: doc/reference/trd106-completion-codes.md

**Draft-Discuss:** tock-dev@googlegroups.com</br>

Abstract
-------------------------------
This document describes the expected behavior of application completion codes
when terminating via the `exit` syscall, as described in [TRD 104][exit-syscall].

1 Introduction
===============================
When an application exits via the [`exit` syscall][exit-syscall], it can specify
a **completion code**, an unsigned 32-bit number which indicates status. This
information can be stored in the kernel and used in management or policy
decisions.

This number is called an "exit status", "exit code", or "result code" on other
platforms.

2 Design Considerations
===============================
When possible, Tock should follow existing conventions and terminology from
other major platforms. This assists in helping the project be more
understandable to newcomers by following the principle of least astonishment.

3 Design
===============================
An application exiting normally via the `exit` syscall MUST use a completion
code of `0` to indicate normal termination. A non-zero completion code SHOULD be
used to indicate abnormal termination. A non-zero completion code MAY be the 

> same value as one of the error codes specified in [TRD 104][error-codes].

Reviewer: It's probably worth making a table here. Something like:

Completion Code | Meaning
------------- | --------
0              | Success
1 - 1024   | Error code as defined in TRD 104
everything else | Not yet defined

Author: Notice the MAY - I specifically don't want to reserve 1-1024. The only required convention is 0 vs. non-zero.

Author: I added a table that I think indicates this

Reviewer 2: IMO, it seems a little confusing to use MAY here. Either we should reserve 1-1024, or we should just say that 1-MAX is not yet defined. Specifying that it may be defined as an error code, or may not, does not seem to have much value

Author: How would you feel about SHOULD? The MAY is to indicate that using 1-1024 for TRD 104 error codes is a good baseline, but an application not following it is still well-formed. I'm thinking from the perspective of an engineer trying to figure out what completion code 12 is, reading this doc, and seeing that it indicates the error code UNINSTALLED by default, providing a valuable hint.

Reviewer 2: I am happy with SHOULD

Author: Thoughts on the new language? I also changed the first sentence, switching the logic from "success implies a 0 completion code" to "0 completion code implies success".

Reviewer: I'm happy with the new language + table.

The kernel MAY treat zero and non-zero completion codes differently.

4 Implementation
===============================

> libtock 1.0 [currently implements][termination] this TRD via the

Reviewer: Shorten this to just `libtock`.

I've been colloquially using the phrase "`libtock-rs` 1.0" to refer to the `libtock-rs` codebase that was designed for Tock 1.0, but we haven't actually named either a 1.0 or a 2.0 release of `libtock`.

Author: Done

itself.

3 Design
===============================
A completion code of `0` passed to the `exit` syscall MUST indicate normal app
termination. A non-zero completion code SHOULD be used to indicate abnormal
termination. This distinction is useful so that a Tock kernel can handle
success/failure cases differently, e.g. by printing error messages,
and so that kernel extensions (such as process exit handlers defined by a board)
or external tools (such as a tool designed to parse the output from a kernel
with *trace\_syscalls* enabled) can match on these two cases.
This behavior also matches the convention for Unix exit codes, such that it
likely matches the expectations for users coming from that domain.

A completion code between `1` and `1024` inclusive SHOULD be the
same value as one of the error codes specified in [TRD 104][error-codes].
This requirement is a SHOULD rather than a MUST because it is useful in the
common case (it allows software to infer something about the cause of an
error that led to an exit, and possibly print a useful message) but also
allows a process to do something else if needed (e.g. for compatibility
with some other standard of exit codes).

> Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
> codes or take actions based on their values besides printing error messages.

Reviewer 3: ```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes or take actions based on their values besides printing error messages, as long as
there is no explicit contract between the kernel and application regarding completion
code interpretation.
```

I like that this does imply that the kernel will not take arbitrary action based solely on the completion codes, but depending on how this is read, it does seem to severely hinder the usefulness of return codes if an app _wants_ do define an explicit contract with the kernel. It might be more clear that this is still an option if we add this phrase explicitly.

Author: I like this, though wouldn't "core kernel" mean that capsules still have the freedom to assume semantic meaning for completion codes?

Reviewer 3: Yes, I think that interpretation is reasonable. Maybe it's worth adding an emphasis to core, as I didn't really realize capsules were deliberately excluded from this statement while reading.

I still think that, given the scheduler and process fault handler be considered part of the core kernel, _allowing_ an explicit contract between an app and the kernel may make sense.

Author: @hudson-ayers, @phil-levis, what are your thoughts on this? I'm fine with the suggested wording here.

Reviewer 2: I am happy with Leon's suggested wording as well

Reviewer 4: Sorry, missed this chain. What's an explicit contract? How would such a contract emerge?

Reviewer 4: @lschuermann @hudson-ayers @kupiakos Just wanted to follow up on this. Would you like me to suggest some text to describe what we talked about today, or Leon, I think you said you'd do that?

Reviewer 3: @phil-levis Sure, didn't get to it until now.

Based on the feedback I suppose it's good to establish my primary motivations behind this change, if only for future record:
- the current wording might be to. restrictive. While downstream users don't technically have to abide to these TRDs, they will the direction of the upstream kernel takes. If we rule out that the kernel can take actions based on app completion codes, we might design internal interfaces that reflect that decision, making it harder for downstream users to implement something like this.
- we _MUST_ ensure that the kernel does not take action on app completion codes as long as there is no guarantee that the kernel knows how to interpret a specific app's completion codes. It is unacceptable for the core kernel to misidentify an app or otherwise misinterpret its completion codes and take action based on this information.

Thus, a more elaborate text might say

```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes. Without a revision of this TRD, the upstream core kernel may not take actions
based on their values besides printing error messages. Future versions of the kernel
may implement mechanisms to take actions based on an app's completion code,
provided that
- the app provides a specification for its completion code space, specifically concerning
  completion codes encoding success and failure cases. It may further specify which types
  of actions the core kernel may enact based on its completion codes.
- the kernel adheres to the app's completion code specification and only takes actions
  allowed by the specification based on the completion code returned. This implies
  that an application with a specification that does not include any instructions about
  actions taken in response to a return code will not be different from an app without such an
  application. The kernel may not implement every action described in this specification.
- the kernel can reliably identify the app and associate it with its completion code specification.
```

This is very elaborate on what a future design could and should look like, but it does not yet fully describe how such a specification must look like and what qualifies as a completion-code specification. It does not necessarily have any benefits for us in the upstream Tock codebase right now, other than being able to design internal interfaces to incorporate such future behavior. Still, it provides downstream users with essential and sensible guidelines on how to implement such a mechanism, which is similar and in the best case compatible with future upstream requirements.

Reviewer 4: How about just trim it down to"

Accordingly, the core kernel MUST NOT assume any semantic meaning for completion codes or take actions based on their values besides printing error messages unless 
  - there is a specification of a particular application's completion code space written in a TRD, and 
  - the kernel can reliably identify that application and associate it with this specification.

Reviewer 2: That works for me

Reviewer 3: @phil-levis Just now catching up on the backlog. Sure, that sounds good and handles all of my concerns. Here's a suggestion which can be applied through GitHub:

```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes or take actions based on their values besides printing error messages
unless

- there is a specification of a particular application's completion code space
  written in a TRD, and

- the kernel can reliably identify that application and associate it with this
  specification.
```

Reviewer 2: I went ahead and applied the suggestion so that outstanding approvals will not get wiped when it is applied

While there are common and conventional uses of certain values, applications
are not required to follow these and may assign their own semantic meanings
to values.

| **Completion Code** | **Meaning**                                   |
| ------------------- | --------------------------------------------- |
| 0                   | Success                                       |
| 1-1024              | SHOULD be a [TRD 104 error code][error-codes] |

> | 1025-`u32::MAX`     | Not defined                                   |

Reviewer 2: ```suggestion
| 1025-`u32::MAX`     | Reserved                                     |
```
